### ISSUE TITLE
Leaked Connection handling

### ISSUE THREAD

**Issue by likaiwalkman** [netty/netty#6794]
### Expected behavior
        connections = right handled requests + abrupt connections 
### Actual behavior
       connections = right handled requests + abrupt connections +1
### Steps to reproduce
       I used netty as http client, then I launched 1000 requests, in ChannelFuture(got from connect)'s handler
       it tells me 975 of them connected, 25 connection failed.
       When I got full and right response from netty-http server, i log a some info.
       I summed right http handling result up, it's totaled as 970.
       In exceptionCaught callback, I log error info, totaled as 4.
       I lost one connection without telling me connection failed or right result came.
### Minimal yet complete reproducer code (or URL to code)
[server and client](https://github.com/likaiwalkman/netty-in-action/blob/master/src/main/java/com/victor/client/api/singleport/)
### Netty version
        4.0.25.Final
### JVM version (e.g. `java -version`)
       1.8 above
### OS version (e.g. `uname -a`)
       Mac OS X

**Comment by normanmaurer** [comment]
Please upgrade to latest netty version and five more informations. Also a reproducer is welcome

**Comment by likaiwalkman** [comment]
this scenerio appears not regularly, when i got a way to make it occur regularly, i will create a git repo including detaled test case.

**Comment by Underbalanced** [comment]
What threading and thread count are you using? I have had some issues that match your numbers when using 1 boss thread only, every 3rd connection would not complete.

**Comment by likaiwalkman** [comment]
I also used one boss thread, so you mean the reason is due to boss thread is not enough?
But I think no matter what happens, no matter what configuration you take, the API should go smoothly.Every request should have an end to tell final phased result -- final success or connection failed or half failed.

**Comment by Underbalanced** [comment]
@likaiwalkman I am unsure if it is related. Would you add an extra thread to the boass thread group to see if it solves the problem?

**Comment by likaiwalkman** [comment]
I got the reason.Increase extra threads won't help.
The reason is due to too many request burst into os exceeding os's max file descriptor limit.
so I got exception saying"**too many open files**".
The code I run can be found in package singleport of my forked repo:
[](https://github.com/likaiwalkman/netty-in-action/blob/master/src/main/java/com/victor/client/api/singleport/HttpClientToSinglePort.java)
The server and the client are both in same package.

**Comment by likaiwalkman** [comment]
The final result are telling when "open too many files"exception is thrown, no event will be watched in the ChannelFuture's callback handler, exception happens when connect method is invoking.
Could all connect events be processed in ChannelFuture's callback handler? @normanmaurer 
I will appreciate it if in that case.

**Comment by likaiwalkman** [comment]
I launched 10000 requests, as I mentioned above, I changed the limit of open files number--**kern.maxfilesperproc=20480**.
I didn't catch exception again where connect method is invoked.
BUT, another thing happens: the **error** log count to 8846, **Ok** log count to 2208, they accumulated to 10854, though the **conns** log count to 9750.
10854 > 9750!!!
error:   printed where exceptionCaught is invoked in HttpClientInboundHandler.
Ok:      printed where channelRead is invoked in HttpClientInboundHandler.
conns: printed where connect is invoked in HttpClientToSinglePort
